File: app/src/main/java/com/digz/cumapapp/navigation/NavigationContract.java
package com.digz.cumapapp.navigation;

import android.bluetooth.BluetoothDevice;

import com.digz.cumapapp.adapter.PlaceAutoCompleteAdapter;
import com.digz.cumapapp.navigation.directioning.DetermineDirection;
import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.maps.model.Polyline;
import com.google.android.gms.maps.model.PolylineOptions;
import com.mapzen.android.lost.api.Status;
import com.mapzen.android.routing.MapzenRouter;
import com.mapzen.helpers.RouteEngine;

public interface NavigationContract {
    interface View{
        void showProgressDialog(String title, String message);
        void dismissProgressDialog();
        void showToast(String message);
        void setStartTripVisibility(int visibility);
        String getTextOfOriginField();
        String getTextOfDestinationField();
        void setErrorOnOriginTextField(String textField);
        void setErrorOnDestinationTextField(String textField);
        void setPlaceAdapterToView(PlaceAutoCompleteAdapter adapter);
        void getResultForLost(Status status);
        void setLeftRightVisibility(int visibility);
        void centerCamera(CameraUpdate center);
        void zoomCamera(CameraUpdate zoom);
        Polyline drawOnMap(PolylineOptions polylineOptions);
        void addMapMarker(MarkerOptions markerOptions);
        MapzenRouter getRouter();
        RouteEngine getRouteEngine();
        DetermineDirection determineDirection();
    }

    interface Presenter{
        void setView(NavigationActivity view);
        void setStartToNull();
        void setEndToNull();
        boolean determineIfOnline();
        void setLocationListener();
        void setupGoogleServices();
        void setUpPlaceAutoCompleteAdapter();
        void route();
        void onDestinationAutocompleteClicked(int position);
        void onStartAutocompleteClicked(int position);
        void createLostApiClient();
        void checkLocationSettings();
        void startTrip();
        void setAdapterBounds(LatLngBounds bounds);
    }
}

File: app/src/main/java/com/digz/cumapapp/navigation/NavigationContract.kt
package com.digz.cumapapp.navigation

import android.location.Location
import com.digz.cumapapp.adapter.PlaceAutoCompleteAdapter
import com.google.android.gms.maps.CameraUpdate
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.gms.maps.model.Polyline
import com.google.android.gms.maps.model.PolylineOptions
import com.mapbox.services.android.navigation.ui.v5.NavigationViewOptions

interface NavigationContract {
    interface View {

        val textOfOriginField: String

        val textOfDestinationField: String

        fun showProgressDialog(title: String, message: String)

        fun dismissProgressDialog()

        fun showToast(message: String)

        fun setStartTripVisibility(visibility: Int)

        fun setErrorOnOriginTextField(error: String?)

        fun setErrorOnDestinationTextField(error: String?)

        fun setPlaceAdapterToView(adapter: PlaceAutoCompleteAdapter)

        fun centerCamera(center: CameraUpdate)

        fun zoomCamera(zoom: CameraUpdate)

        fun drawOnMap(polylineOptions: PolylineOptions): Polyline

        fun setStartAutoCompleteText()

        fun addMapMarker(markerOptions: MarkerOptions)

        fun startNavigation(options:NavigationViewOptions)

    }

    interface Presenter {
        fun onAttach(view: NavigationActivity)

        fun setStartToNull()

        fun setEndToNull()

        fun setStart(location:Location)

        fun determineIfOnline(): Boolean

        fun setLocationListener()

        fun setupGoogleServices()

        fun setUpPlaceAutoCompleteAdapter()

        fun route()

        fun onDestinationAutocompleteClicked(position: Int)

        fun onStartAutocompleteClicked(position: Int)

        fun checkLocationSettings()

        fun startTrip()

        fun setAdapterBounds(bounds: LatLngBounds)
    }
}
